<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.01//EN" "http://www.w3.org/TR/html4/strict.dtd">
<html>
<head>
<meta content="text/html; charset=ISO-8859-1" http-equiv="content-type">
<title>Reference | GBlocks</title></head>
<body>
<big style="font-weight: bold; font-family: Courier New,Courier,monospace;"><big><big><span style="color: rgb(102, 102, 102);">Reference Armadillo v1.1 </span></big></big></big><br><br>
<table style="text-align: left; width: 800px; height: 300px;" border="0" cellpadding="2" cellspacing="2">
<tbody><tr><td></td><td style="width: 215px;">Name</td><td style="width: 1004px;"><big><span style="font-weight: bold;">GBlocks</span></big></td></tr>
<tr><td></td><td>View</td><td><span style="font-family: Courier New,Courier,monospace;"><img src="GBlocks.png"></img></span></td></tr>
<tr><td></td><td style="width: 215px; vertical-align: top;">Publication(s)</td><td style="width: 1004px; align:justify ">Talavera, G., and Castresana, J. (2007). Improvement of phylogenies after removing divergent and ambiguously aligned blocks from protein sequence alignments. Systematic Biology 56, 564-577. 

Castresana, J. (2000). Selection of conserved blocks from multiple alignments for their use in phylogenetic analysis. Molecular Biology and Evolution 17, 540-552.<br><br></td></tr>
<tr><td></td><td style="width: 215px; vertical-align: top;">Website</td><td style="width: 1004px; align:justify "><b><a href='http://molevol.cmima.csic.es/castresana/Gblocks.html'>http://molevol.cmima.csic.es/castresana/Gblocks.html</b></a><br><br></td></tr>
<tr><td></td><td style="width: 215px; vertical-align: top;">Description</td><td style="width: 1004px; align:justify ">Version 0.91b, January 2002 Copyright Â© Jose Castresana

Eliminates poorly aligned positions and divergent regions of an alignment of DNA or protein sequences. These positions may not be homologous or may have been saturated by multiple substitutions and it is convenient to eliminate them prior to phylogenetic analysis. Gblocks selects blocks in a similar way as it is usually done by hand but following a reproducible set of conditions. The selected blocks must fulfill certain requirements with respect to the lack of large segments of contiguous nonconserved positions, lack or low density of gap positions and high conservation of flanking positions, making the final alignment more suitable for phylogenetic analysis.<br><br></td></tr>
<tr><td></td><td style="width: 215px; vertical-align: top;">Input (<strong>1</strong>)</td><td style="width: 215px; vertical-align: top;">
Alignment<br></td></tr><tr><td></td><td style="width: 215px; vertical-align: top;">Output (<strong>3</strong>)</td><td style="width: 215px; vertical-align: top;">
Results<br>
Alignment<br>
OutputText<br>
</td></tr>
<tr><td></td>
<tr><td></td><td style="width: 215px; vertical-align: top;">Keywords (<strong>1</strong>)</td><td style="width: 215px; vertical-align: top;">
Sequences, Column, Selection &nbsp</td></tr>
<tr><td></td><td style="width: 215px; vertical-align: top;">Sample workflow</td><td style="width: 1004px; align:justify "><a href='Gblocks.properties.db'>Gblocks.properties.db</a><br><br></td></tr>
</tbody></table><br></body></html>
